File: sdk/cpp/tests/test_query.cpp
/**
 * Tests for the0::query module
 */

#include <doctest/doctest.h>
#include "helpers/test_helpers.hpp"
#include "../the0/query.h"
#include "../the0/state.h"

using namespace the0;
using namespace the0_test;
using json = nlohmann::json;

TEST_CASE("query::is_query_mode") {
    SUBCASE("returns false when QUERY_PATH not set") {
        unsetenv("QUERY_PATH");

        CHECK_FALSE(query::is_query_mode());
    }

    SUBCASE("returns true when QUERY_PATH is set") {
        SetEnv env("QUERY_PATH", "/portfolio");

        CHECK(query::is_query_mode());
    }

    SUBCASE("returns false when QUERY_PATH is empty") {
        SetEnv env("QUERY_PATH", "");

        CHECK_FALSE(query::is_query_mode());
    }
}

TEST_CASE("query::get_params") {
    SUBCASE("returns empty map by default") {
        auto params = query::get_params();

        CHECK(params.empty());
    }
}

TEST_CASE("query::get_config") {
    SUBCASE("returns empty object by default") {
        auto config = query::get_config();

        CHECK(config.is_object());
    }
}

TEST_CASE("query::handler registration") {
    SUBCASE("registers handler without throwing") {
        CHECK_NOTHROW(
            query::handler("/test", [](const query::Request& req) {
                return json{{"test", true}};
            })
        );
    }

    SUBCASE("can register multiple handlers") {
        CHECK_NOTHROW([&]() {
            query::handler("/one", [](const query::Request& req) {
                return json{{"id", 1}};
            });
            query::handler("/two", [](const query::Request& req) {
                return json{{"id", 2}};
            });
        }());
    }
}

TEST_CASE("query::Request") {
    SUBCASE("get returns value for existing key") {
        std::map<std::string, std::string> params = {{"symbol", "BTC/USD"}};
        query::Request req("/test", params);

        CHECK(req.get("symbol") == "BTC/USD");
    }

    SUBCASE("get returns empty for missing key") {
        std::map<std::string, std::string> params;
        query::Request req("/test", params);

        CHECK(req.get("missing") == "");
    }

    SUBCASE("get with default returns value for existing key") {
        std::map<std::string, std::string> params = {{"symbol", "BTC/USD"}};
        query::Request req("/test", params);

        CHECK(req.get("symbol", "default") == "BTC/USD");
    }

    SUBCASE("get with default returns default for missing key") {
        std::map<std::string, std::string> params;
        query::Request req("/test", params);

        CHECK(req.get("missing", "default") == "default");
    }

    SUBCASE("has returns true for existing key") {
        std::map<std::string, std::string> params = {{"symbol", "BTC/USD"}};
        query::Request req("/test", params);

        CHECK(req.has("symbol"));
    }

    SUBCASE("has returns false for missing key") {
        std::map<std::string, std::string> params;
        query::Request req("/test", params);

        CHECK_FALSE(req.has("missing"));
    }

    SUBCASE("path is accessible") {
        std::map<std::string, std::string> params;
        query::Request req("/portfolio", params);

        CHECK(req.path == "/portfolio");
    }
}

TEST_CASE("state read-only enforcement in query mode") {
    // Set up temp directory for state
    auto temp_dir = std::filesystem::temp_directory_path() / ("the0_query_test_" + std::to_string(rand()));
    std::filesystem::create_directories(temp_dir);
    setenv("STATE_DIR", temp_dir.c_str(), 1);

    SUBCASE("state::set throws in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        CHECK_THROWS_AS(
            state::set("key", json{{"test", true}}),
            state::ReadOnlyStateError
        );
    }

    SUBCASE("state::remove throws in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        CHECK_THROWS_AS(
            state::remove("key"),
            state::ReadOnlyStateError
        );
    }

    SUBCASE("state::clear throws in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        CHECK_THROWS_AS(
            state::clear(),
            state::ReadOnlyStateError
        );
    }

    SUBCASE("state::get allowed in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        // Should not throw
        auto result = state::get("nonexistent");
        CHECK_FALSE(result.has_value());
    }

    SUBCASE("state::get_or allowed in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        // Should not throw
        auto result = state::get_or("nonexistent", json("default"));
        CHECK(result == json("default"));
    }

    SUBCASE("state::list allowed in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        // Should not throw
        auto result = state::list();
        CHECK(result.empty());
    }

    SUBCASE("state::exists allowed in query mode") {
        SetEnv queryEnv("QUERY_PATH", "/test");

        // Should not throw
        bool result = state::exists("nonexistent");
        CHECK_FALSE(result);
    }

    // Cleanup
    unsetenv("STATE_DIR");
    std::filesystem::remove_all(temp_dir);
}

File: sdk/rust/tests/test_query.rs
//! Tests for the0_sdk::query module

mod common;

use common::EnvGuard;
use the0_sdk::query;
use the0_sdk::state;

#[test]
fn is_query_mode_returns_false_when_not_set() {
    let mut env = EnvGuard::new();
    env.remove("QUERY_PATH");

    assert!(!query::is_query_mode());
}

#[test]
fn is_query_mode_returns_true_when_set() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");

    assert!(query::is_query_mode());
}

#[test]
fn get_params_returns_empty_by_default() {
    let _env = EnvGuard::new();
    let params = query::get_params();
    assert!(params.is_empty());
}

#[test]
fn get_config_returns_empty_object_by_default() {
    let _env = EnvGuard::new();
    let config = query::get_config();
    assert!(config.is_object());
}

#[test]
fn handler_registers_without_calling() {
    let _env = EnvGuard::new();

    // Just verify that handler registration works without panicking
    // We can't easily test the actual execution without starting the server
    query::handler("/test", |_req| {
        serde_json::json!({"test": true})
    });
}

#[test]
fn query_request_get_returns_value() {
    let mut params = std::collections::HashMap::new();
    params.insert("symbol".to_string(), "BTC/USD".to_string());

    let req = query::QueryRequest::new("/test", params);

    assert_eq!(req.get("symbol"), Some("BTC/USD".to_string()));
    assert_eq!(req.get("missing"), None);
}

#[test]
fn query_request_get_or_returns_default() {
    let mut params = std::collections::HashMap::new();
    params.insert("symbol".to_string(), "BTC/USD".to_string());

    let req = query::QueryRequest::new("/test", params);

    assert_eq!(req.get_or("symbol", "default"), "BTC/USD");
    assert_eq!(req.get_or("missing", "default"), "default");
}

#[test]
fn state_set_throws_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state");

    let result = state::set("test_key", &serde_json::json!({"test": true}));

    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(matches!(err, state::StateError::ReadOnly(_)));
}

#[test]
fn state_delete_returns_false_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state");

    // delete returns false in query mode (can't modify state)
    let result = state::delete("test_key");
    assert!(!result);
}

#[test]
fn state_clear_throws_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state");

    let result = state::clear();

    assert!(result.is_err());
    let err = result.unwrap_err();
    assert!(matches!(err, state::StateError::ReadOnly(_)));
}

#[test]
fn state_get_allowed_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state-nonexistent");

    // get should work even in query mode (read-only operation)
    // Returns Err for non-existent file, but NOT a ReadOnly error
    let result: Result<String, _> = state::get("nonexistent");
    assert!(result.is_err());
    // Verify it's an IO error (file not found), not a ReadOnly error
    let err = result.unwrap_err();
    assert!(!matches!(err, state::StateError::ReadOnly(_)));
}

#[test]
fn state_get_or_allowed_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state-nonexistent");

    // get_or should work even in query mode (read-only operation)
    let result: String = state::get_or("nonexistent", "default".to_string());
    assert_eq!(result, "default");
}

#[test]
fn state_list_allowed_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state-nonexistent");

    // list should work even in query mode (read-only operation)
    let result = state::list();
    assert!(result.is_empty());
}

#[test]
fn state_exists_allowed_in_query_mode() {
    let mut env = EnvGuard::new();
    env.set("QUERY_PATH", "/portfolio");
    env.set("STATE_DIR", "/tmp/the0-test-state-nonexistent");

    // exists should work even in query mode (read-only operation)
    let result = state::exists("nonexistent");
    assert!(!result);
}
